src/crypto-providers/signingValidation.ts

```diff
@@ -0,0 +1,232 @@
+    (withdrawal) => withdrawal.stakeCredential.type === StakeCredentialType.ADDR_KEY_HASH,
+  )
+  if (!certificatesOk || !withdrawalsOk) {
+    throw Error(Errors.ScriptStakeCredentialInOrdinaryTx)
+  }
+}
+
+const _countWithdrawalAndCertificateWitnesses = (unsignedTxParsed: _UnsignedTxParsed) => {
+  let numStakeWitnesses = unsignedTxParsed.withdrawals.length
+  let numPoolColdWitnesses = 0
+  unsignedTxParsed.certificates.forEach((cert) => {
+    switch (cert.type) {
+      case TxCertificateKeys.STAKING_KEY_REGISTRATION:
+      case TxCertificateKeys.STAKING_KEY_DEREGISTRATION:
+      case TxCertificateKeys.DELEGATION:
+        numStakeWitnesses += 1
+        break
+
+      case TxCertificateKeys.STAKEPOOL_RETIREMENT:
+        numPoolColdWitnesses += 1
+        break
+
+      default:
+        break
+    }
+  })
+  return { numStakeWitnesses, numPoolColdWitnesses }
+}
+
+const validateOrdinaryTx = (unsignedTxParsed: _UnsignedTxParsed) => {
+  _validateNonMultisigTx(unsignedTxParsed)
+
+  if (unsignedTxParsed.inputs.length === 0) {
+    throw Error(Errors.MissingInputError)
+  }
+}
+
+// TODO we don't count unique witnesses
+const validateOrdinaryWitnesses = (params: SigningParameters) => {
+  const {
+    stakeSigningFiles, poolColdSigningFiles, mintSigningFiles, multisigSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  const {
+    numStakeWitnesses, numPoolColdWitnesses,
+  } = _countWithdrawalAndCertificateWitnesses(params.unsignedTxParsed)
+  const numMintWitnesses = params.unsignedTxParsed.mint?.length || 0
+
+  if (stakeSigningFiles.length > numStakeWitnesses) {
+    throw Error(Errors.TooManyStakeSigningFilesError)
+  }
+  if (poolColdSigningFiles.length > numPoolColdWitnesses) {
+    throw Error(Errors.TooManyPoolColdSigningFilesError)
+  }
+  if (mintSigningFiles.length > numMintWitnesses) {
+    throw Error(Errors.TooManyMintSigningFilesError)
+  }
+  if (multisigSigningFiles.length > 0) {
+    throw Error(Errors.TooManyMultisigSigningFilesError)
+  }
+}
+
+const validateOrdinarySigning = (params: SigningParameters) => {
+  // these checks should be performed only with createSignedTx call (when the signing file set is
+  // expected to be complete) on top of validateOrdinaryWitnesses
+  const {
+    paymentSigningFiles, stakeSigningFiles, poolColdSigningFiles, mintSigningFiles,
+  } = filterSigningFiles(params.hwSigningFileData)
+
+  if (paymentSigningFiles.length === 0) {
+    throw Error(Errors.MissingPaymentSigningFileError)
+  }
+
+  const {
+    numStakeWitnesses, numPoolColdWitnesses,
+  } = _countWithdrawalAndCertificateWitnesses(params.unsignedTxParsed)
+  const numMintWitnesses = params.unsignedTxParsed.mint?.length || 0
+
+  if (numStakeWitnesses > 0 && (stakeSigningFiles.length === 0)) {
+    throw Error(Errors.MissingStakeSigningFileError)
+  }
+  if (numPoolColdWitnesses > 0 && (poolColdSigningFiles.length === 0)) {
+    throw Error(Errors.MissingPoolColdSigningFileError)
+  }
```

Comment:
So here we are validating whether we have some stake/poolCold witnesses when they are required, but as opposed to witnessing we are not checking whether there are too many of them? What's the reason?

Reply:
Okay I just realized that when this signing validation is called we also call the `validateOrdinaryWitnesses`, I think it would be best to call it from inside this function and not call these two separately in the `validateSigning` function.

Reply:
Fixed.